c7n/resources/vpc.py

```diff
@@ -2708,3 +2708,80 @@ class resource_type(query.TypeInfo):
+                      op: eq
+                      value: not-null
+                    - type: value
+                      key: NatGatewayInCrossAvailabilityZone
+                      op: eq
+                      value: true
+                actions:
+                  - notify
+    """
+    schema = type_schema('cross-az-nat-gateway-route')
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client('ec2')
+        # dump of all subnets to avoid multiple API calls
+        all_subnets = client.describe_subnets()
+        all_nat_gws = client.describe_nat_gateways()
+        subnets_az_map = {item["SubnetId"]: item["AvailabilityZone"] for item in all_subnets["Subnets"]}
+        nat_gws_subnet_map = {item['NatGatewayId']: item["SubnetId"] for item in all_nat_gws["NatGateways"]}
+        all_nat_gw_az = [subnets_az_map[item["SubnetId"]] for item in
+                         all_nat_gws["NatGateways"]]  # To make sure NAT Gateway is available in a given AZ
+
+        for res in resources:
+            res['NatGatewayAvailabilityZone'] = {}
+            for item in res["Routes"]:
+                if item.get("NatGatewayId") and item.get("State") == "active":
+                    # Get subnet associations
+                    subnet_associations = [item["SubnetId"] for item in res["Associations"] if item.get("SubnetId")]
+                    # Get NAT Gateway availability zone
+                    nat_gw_az = subnets_az_map[nat_gws_subnet_map[item.get("NatGatewayId")]]
+                    res['NatGatewayAvailabilityZone'][item.get("NatGatewayId")] = nat_gw_az
+                    # Get subnet AZs
+                    if subnet_associations:
+                        cross_nat_az = list()  # To store cross NAT AZ status
+                        for item1 in res["Associations"]:
+                            if item1.get("SubnetId"):
+                                item1['SubnetAvailabilityZone'] = subnets_az_map[item1["SubnetId"]]
+                                if subnets_az_map[item1["SubnetId"]] in all_nat_gw_az:
+                                    item1['NatGatewayAvailableInSubnetAvailabilityZone'] = True
+                                else:
+                                    item1['NatGatewayAvailableInSubnetAvailabilityZone'] = False
+                                # check if Associated Subnet AZ is same as NAT GW AZ
+                                if subnets_az_map[item1["SubnetId"]] != nat_gw_az:
+                                    cross_nat_az.append('Y')
+                                else:
+                                    cross_nat_az.append('N')
+                        if 'Y' in cross_nat_az:
+                            res['NatGatewayInCrossAvailabilityZone'] = True
+                        else:
+                            res['NatGatewayInCrossAvailabilityZone'] = False
+        return resources
```

Comment:
@vvishwakarma-intuit This is confusing because you're just returning ALL of the resources even though we're supposed to be filtering them.

Reply:
Returning the filtered resources as suggested.

Reply:
Returning the filtered resources as suggested.